install.c

```diff
@@ -1,28 +1,71 @@
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
 #include "substdio.h"
 #include "strerr.h"
+#include "fifo.h"
 #include "error.h"
 #include "open.h"
-#include "readwrite.h"
 #include "exit.h"
+#include "stralloc.h"
+#include "sgetopt.h"
 
 extern void hier();
 
 #define FATAL "install: fatal: "
 
 int fdsourcedir = -1;
+uid_t my_uid;
+gid_t my_gid;
+char *destdir = 0;
+stralloc tmpdir = { 0 };
+stralloc dirbuf = { 0 };
+
+int
+r_mkdir(home, mode)
+	char           *home;
+	int             mode;
+{
+	char           *ptr;
+	int             i;
+
+	if (!stralloc_copys(&dirbuf, home))
+		strerr_die2sys(111, FATAL, "out of memory: ");
+	if (!stralloc_0(&dirbuf))
+		strerr_die2sys(111, FATAL, "out of memory: ");
+	for (ptr = dirbuf.s + 1;*ptr;ptr++) {
+		if (*ptr == '/') {
+			*ptr = 0;
+			if (access(dirbuf.s, F_OK) && (i = mkdir(dirbuf.s, mode)) == -1)
```

Comment:
there is a race condition here: if someone creates the directory between the access() and the mkdir() the function will fail. You need to check errno == error_exist after the mkdir().